### ISSUE TITLE
InvalidIRError when ndims(count) or ndims(prob) is >= 3 units smaller than ndims(A)

### ISSUE THREAD

**Issue by simsurace** [JuliaGPU/BinomialGPU.jl#3]
Minimal example:
```
using CUDA, BinomialGPU

A = CUDA.zeros(Int, 2, 2, 2, 2)
ns = CUDA.fill(128, (2, 2))
ps = CUDA.rand(2)

rand_binomial!(A, count = ns, prob = ps)

**Comment by simsurace** [comment]
<details>
<summary>Error message</summary>

``` 
InvalidIRError: compiling kernel BTRS_full(CuDeviceArray{Int64, 4, 1}, CuDeviceMatrix{Int64, 1}, CuDeviceVector{Float32, 1}, CuDeviceVector{NTuple{4, UInt32}, 1}) resulted in invalid LLVM IR
Reason: unsupported call through a literal pointer (call to )
Stacktrace:
 [1] Array
   @ boot.jl:448
 [2] Array
   @ boot.jl:457
 [3] similar
   @ abstractarray.jl:785
 [4] similar
   @ abstractarray.jl:784
 [5] _array_for
   @ array.jl:670
 [6] getindex
   @ range.jl:313
 [7] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:42
Reason: unsupported call to the Julia runtime (call to jl_f__apply_iterate)
Stacktrace:
 [1] getindex
   @ range.jl:313
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:42
Reason: unsupported dynamic function invocation (call to CartesianIndex)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:42
Reason: unsupported dynamic function invocation (call to getindex)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:42
Reason: unsupported dynamic function invocation (call to <=)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:49
Reason: unsupported dynamic function invocation (call to >=)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:52
Reason: unsupported dynamic function invocation (call to <)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:61
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:61
Reason: unsupported dynamic function invocation (call to convert)
Stacktrace:
 [1] setindex!
   @ ~/.julia/packages/CUDA/qEV3Y/src/device/array.jl:103
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:61
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:67
Reason: unsupported dynamic function invocation (call to <)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:67
Reason: unsupported dynamic function invocation (call to -)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:68
Reason: unsupported dynamic function invocation (call to log)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:68
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:72
Reason: unsupported dynamic function invocation (call to ceil)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:72
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:73
Reason: unsupported dynamic function invocation (call to >)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:74
Reason: unsupported dynamic function invocation (call to >)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:83
Reason: unsupported dynamic function invocation (call to -)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:84
Reason: unsupported dynamic function invocation (call to -)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:86
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:86
Reason: unsupported dynamic function invocation (call to -)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:87
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:87
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:89
Reason: unsupported dynamic function invocation (call to sqrt)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:89
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:90
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:90
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:91
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] +
   @ operators.jl:560
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:91
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:92
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:92
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:93
Reason: unsupported dynamic function invocation (call to -)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:93
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:95
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:95
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:95
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:96
Reason: unsupported dynamic function invocation (call to floor)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:96
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:104
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:104
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:104
Reason: unsupported dynamic function invocation (call to floor)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:104
Reason: unsupported dynamic function invocation (call to <=)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:106
Reason: unsupported dynamic function invocation (call to convert)
Stacktrace:
 [1] setindex!
   @ ~/.julia/packages/CUDA/qEV3Y/src/device/array.jl:103
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:107
Reason: unsupported dynamic function invocation (call to <)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:111
Reason: unsupported dynamic function invocation (call to >)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:111
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:115
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:115
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:115
Reason: unsupported dynamic function invocation (call to log)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:115
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to -)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to *)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to /)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to log)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to stirling_approx_tail)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to +)
Stacktrace:
 [1] +
   @ operators.jl:560
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to afoldl)
Stacktrace:
 [1] +
   @ operators.jl:560
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:116
Reason: unsupported dynamic function invocation (call to <=)
Stacktrace:
 [1] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:120
Reason: unsupported dynamic function invocation (call to convert)
Stacktrace:
 [1] setindex!
   @ ~/.julia/packages/CUDA/qEV3Y/src/device/array.jl:103
 [2] kernel_BTRS!
   @ ~/.julia/dev/BinomialGPU/src/kernels.jl:121
in eval at base/boot.jl:360 
in top-level scope at test.jl:57
in  at dev/BinomialGPU/src/rand_binomial.jl:9
in #rand_binomial!#1 at dev/BinomialGPU/src/rand_binomial.jl:9 
in rand_binomial!##kw at dev/BinomialGPU/src/rand_binomial.jl:30 
in #rand_binomial!#8 at dev/BinomialGPU/src/rand_binomial.jl:30 
in rand_binom! at dev/BinomialGPU/src/rand_binomial.jl:56
in macro expansion at CUDA/qEV3Y/src/compiler/execution.jl:102 
in  at CUDA/qEV3Y/src/compiler/execution.jl:288
in var"#cufunction#221" at CUDA/qEV3Y/src/compiler/execution.jl:294
in cached_compilation at GPUCompiler/XwWPj/src/cache.jl
in cached_compilation at dev/BinomialGPU/src/kernels.jl:36 
in check_cache at GPUCompiler/XwWPj/src/cache.jl:44 
in cufunction_compile at CUDA/qEV3Y/src/compiler/execution.jl:306
in  at GPUCompiler/XwWPj/src/utils.jl:60
in var"#emit_asm#103" at GPUCompiler/XwWPj/src/utils.jl:62
in macro expansion at GPUCompiler/XwWPj/src/driver.jl:286 
in macro expansion at TimerOutputs/4QAIk/src/TimerOutput.jl:206 
in macro expansion at GPUCompiler/XwWPj/src/driver.jl:288 
in check_ir at GPUCompiler/XwWPj/src/validation.jl:123
```
</details>

**Comment by simsurace** [comment]
This is why the build is currently failing. If this issue is resolved, all current tests (as of version 0.2.3) should pass.

**Comment by simsurace** [comment]
Fixed by commit 0eedb51fce154740b17ab64fcef627df4ddc2806

**Comment by simsurace** [comment]
Fixed by commit 0eedb51
